# 🐋 循环添加元素的速度问题

背景：在数据预处理或特征提取中，常常对多个文件执行特征提取操作，最终将提取的特征合并为一个数组并保存为`.npy`文件。

一个小问题是：怎样添加列表能达到最快速度呢？

下面简单对比几种元素添加方式

## 壹丨对比

In [2]:
import time
import numpy as np

### 1. 使用list的append方法添加，再转numpy

In [7]:
def run_list():
    ss = time.time()
    res = []
    for i in range(1000000):
        res.append(i)
    res = np.asarray(res)
    print(f'用时 = {time.time() - ss}')

run_list()

用时 = 0.10399866104125977


### 2. 使用list的+=方法添加，再转numpy

In [10]:
def run_plus():
    ss = time.time()
    res = []
    for i in range(1000000):
        res += [i]
    res = np.asarray(res)
    print(f'用时 = {time.time() - ss}')

run_plus()

用时 = 0.1379992961883545


### 3. 推导式添加list，再转numpy

In [12]:
def run_add():
    ss = time.time()
    res = np.asarray([i for i in range(1000000)])
    print(f'用时 = {time.time() - ss}')

run_plus()

用时 = 0.0859994888305664


### 4. 使用numpy.append方法添加元素

In [13]:
def run_np():
    ss = time.time()
    res = np.array([])
    for i in range(100000):
        res = np.append(res, i)
    print(f'用时 = {time.time() - ss}')

run_np()

用时 = 2.1151256561279297


## 贰丨总结

| 方法             | 时间                |
| ---------------- | ------------------- |
| list推导式       | 0.0859994888305664  |
| list的`append()` | 0.10399866104125977 |
| list的`+=`       | 0.1379992961883545  |
| `numpy.append()` | 2.1151256561279297  |